device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/amqps/AmqpsIotHubConnection.java

```diff
@@ -211,13 +212,17 @@ public void open(Queue<DeviceClientConfig> deviceClientConfigs) throws Transport
 
                 if (this.savedException != null)
                 {
+                    amqpsSessionManager.closeNow();
+                    executorServicesCleanup();
                     // Codes_SRS_AMQPSIOTHUBCONNECTION_34_062: [If, after attempting to open the connection, this
                     // object has a saved exception, this function shall throw that saved exception.]
                     throw this.savedException;
                 }
 
                 if (!this.amqpsSessionManager.isAuthenticationOpened() || !this.amqpsSessionManager.areAllLinksOpen() || this.state != IotHubConnectionStatus.CONNECTED)
                 {
+                    amqpsSessionManager.closeNow();
```

Comment:
>amqpsSessionManager [](start = 20, length = 19)

unit test updates?

Reply:
otherwise, this looks good to me

---
In reply to: [251634930](https://github.com/Azure/azure-iot-sdk-java/pull/423#discussion_r251634930) [](ancestors = 251634930)